### ISSUE TITLE
Reproducibility question: commented-out stage-two loss and empty UAEO index handling

### ISSUE THREAD

**Issue by YuKDseele** [JiayinXu5499/PSIC#1]
Hello authors,

## Summary

We are trying to reproduce the reported PSIC results from the public repository. During a code review of the original training implementation, we noticed two details that may affect reproducibility:

## Potential implementation issues

1. In the public stage-two training script, the code constructing the final training loss appears to be commented out. In particular, the assignment that combines the rate-distortion terms and the UAEO/CLIP term appears disabled. Could you confirm whether this is intentional, or whether the released script is missing a line or patch?

2. In the UAEO implementation, the selected clean and noisy index sets can be empty (`c_idx` or `n_idx`). A mean cross-entropy computed on an empty subset can produce `NaN`, which can then propagate through training. Could you clarify how the official implementation is intended to handle this case? Should the selection be constrained so that both groups are non-empty, or is there another fallback used in the reported experiments?

## Reproduction results

We then reran the relevant stage-one and stage-two experiments. The following are our evaluation results at the indicated checkpoints:

| Run | PSNR encoded | PSNR protected | BPP | Protected retrieval i2t | Protected retrieval t2i | ASR i2t | ASR t2i | Mean ASR |
|---|---:|---:|---:|---:|---:|---:|---:|---:|
| Original UAEO stage 1 | 24.8729 | 24.8728 | 0.135924 | 0.242368 | 0.177135 | 0.048657 | 0.063565 | 0.056111 |
| Global original UAEO stage 2, epoch 100 | 24.9610 | 24.9212 | 0.135924 | 0.239525 | 0.170436 | 0.165698 | 0.211808 | 0.188753 |

These results differ substantially from the expected behavior and from the reported results. We may be misunderstanding the intended training path, loss definition, checkpoint initialization, or hyperparameter settings.

## Questions

Could you please clarify:

- Which exact commit or training script was used for the reported results?
- Is the commented-out final loss assignment an omission in the public code?
- How are empty `c_idx`/`n_idx` cases handled in the official experiments?
- Are there additional patches, configuration files, preprocessing choices, or checkpoint settings required for reproduction?
- Could you provide a verified runnable script or the exact patch used for the reported results?

## Available artifacts

We can provide our code diff, full commands, logs, and the complete metric table if useful. Thank you for your help.

**Comment by JiayinXu5499** [comment]
Dear researchers,

Thank you very much for your careful reproduction and detailed feedback. We have carefully checked your questions and experimental results, and here are our official clarifications.

First, the commented-out loss combination code in the stage-two script is indeed an oversight during code release. We have now fixed this bug and restored the complete joint loss of rate-distortion and UAEO/CLIP terms.

Second, regarding the empty index issue: empty `c_idx` or `n_idx` rarely occurred in our original experiments due to the large batch size. To avoid potential NaN problems, we have added a reasonable fallback strategy. When the noisy index set is empty, the model uses only the clean sample loss for stable training.

We also notice the inconsistency between your reproduced results and the paper. According to our paper’s two-stage design, Stage 1 focuses on visual compression quality, while Stage 2 sacrifices partial visual quality to improve the ASR(privacy defense performance). The abnormal PSNR rise and insufficient ASR growth in your Stage 2 results are caused by the incomplete loss function in the original public code.

In addition, we have updated the official running commands for stage-two training. We apologize for the confusion and adverse impact on your reproduction.

All configurations and training settings described in our paper are correct. After you pull the latest code and use the updated stage-two command, the experimental trade-off performance will match the results reported in our paper.

If problems still exist after re-running, you are welcome to provide your logs and running commands for further investigation.

Best regards,
Authors

**Comment by rsr1123** [comment]
> Dear researchers,
> 
> Thank you very much for your careful reproduction and detailed feedback. We have carefully checked your questions and experimental results, and here are our official clarifications.
> 
> First, the commented-out loss combination code in the stage-two script is indeed an oversight during code release. We have now fixed this bug and restored the complete joint loss of rate-distortion and UAEO/CLIP terms.
> 
> Second, regarding the empty index issue: empty `c_idx` or `n_idx` rarely occurred in our original experiments due to the large batch size. To avoid potential NaN problems, we have added a reasonable fallback strategy. When the noisy index set is empty, the model uses only the clean sample loss for stable training.
> 
> We also notice the inconsistency between your reproduced results and the paper. According to our paper’s two-stage design, Stage 1 focuses on visual compression quality, while Stage 2 sacrifices partial visual quality to improve the ASR(privacy defense performance). The abnormal PSNR rise and insufficient ASR growth in your Stage 2 results are caused by the incomplete loss function in the original public code.
> 
> In addition, we have updated the official running commands for stage-two training. We apologize for the confusion and adverse impact on your reproduction.
> 
> All configurations and training settings described in our paper are correct. After you pull the latest code and use the updated stage-two command, the experimental trade-off performance will match the results reported in our paper.
> 
> If problems still exist after re-running, you are welcome to provide your logs and running commands for further investigation.
> 
> Best regards, Authors

Thank you for sharing the corrected Stage 2 code. I noticed that `n_loss` is assigned twice consecutively:

```python
n_loss = 0.5 * (
    clip_loss(n_per_output, output_text, 0, i_label)
    + clip_loss(n_output_text, per_output1, 0, t_label)
).mean()

n_loss = 0.5 * (
    clip_loss(n_per_output, output_text, 0, n_idx)
    + clip_loss(n_output_text, per_output1, 0, n_idx)
).mean()
```

The second assignment overwrites the first. Since `CLIP_Loss` uses ordinary cross-entropy for `target=0`, minimizing the second loss appears to encourage the protected images and texts to match their original paired indices, rather than the low-relevance targets selected by `i_label` and `t_label`.

I noticed that the second assignment was commented out in the earlier version, so I wanted to check which formulation we should use for reproduction. I may be missing some context—could you help me understand the role of using `n_idx` here? Thank you again for your help!

Thank you for your clarification!

**Comment by YuKDseele** [comment]
> Dear researchers,
> 
> Thank you very much for your careful reproduction and detailed feedback. We have carefully checked your questions and experimental results, and here are our official clarifications.
> 
> First, the commented-out loss combination code in the stage-two script is indeed an oversight during code release. We have now fixed this bug and restored the complete joint loss of rate-distortion and UAEO/CLIP terms.
> 
> Second, regarding the empty index issue: empty `c_idx` or `n_idx` rarely occurred in our original experiments due to the large batch size. To avoid potential NaN problems, we have added a reasonable fallback strategy. When the noisy index set is empty, the model uses only the clean sample loss for stable training.
> 
> We also notice the inconsistency between your reproduced results and the paper. According to our paper’s two-stage design, Stage 1 focuses on visual compression quality, while Stage 2 sacrifices partial visual quality to improve the ASR(privacy defense performance). The abnormal PSNR rise and insufficient ASR growth in your Stage 2 results are caused by the incomplete loss function in the original public code.
> 
> In addition, we have updated the official running commands for stage-two training. We apologize for the confusion and adverse impact on your reproduction.
> 
> All configurations and training settings described in our paper are correct. After you pull the latest code and use the updated stage-two command, the experimental trade-off performance will match the results reported in our paper.
> 
> If problems still exist after re-running, you are welcome to provide your logs and running commands for further investigation.
> 
> Best regards, Authors

Thank you for clarifying our earlier questions—it has been very helpful. As we continue with the reproduction, there are a couple of additional details we are still unsure about:

1. We found the configuration for Q1, but are not quite sure about the settings for Q2–Q4. Would you mind sharing the values of `lambda`, `lambda_beta0`, `lambda_beta1`, and `lambda_clip` used in each training stage for these quality levels? Does the training schedule remain the same as for Q1?

2. For the AE-Hyperprior (Ballé2018) baseline reported in the paper, could you also share the rate–distortion λ values, batch size, number of training epochs or iterations, and learning-rate schedule?

Apologies if we have overlooked these details. Any additional guidance or configuration files would be greatly appreciated and would help us follow your experimental setup more closely.

Thank you again for your time and help!
